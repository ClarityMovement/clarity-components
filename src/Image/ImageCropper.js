import React from 'react';
import T from 'prop-types';
import { compose, withHandlers } from 'recompose';

import AvatarEditor from 'react-avatar-editor';
import Slider from 'material-ui/Slider';

import withStates from '../hocs/withStates';

const staticProps = {
  slider: {
    min: 1,
    max: 2,
  },
  cropper: {
    crossOrigin: 'anonymous',
    border: 34,
    color: [222, 222, 222, 0.6],
    rotate: 0,
  },
};

const styles = {
  slider: { margin: '5px 0' },
};

const ImageCropperRenderer = ({ image, scale, refEditor, onScaleChange }) => (
  <div>
    <AvatarEditor
      {...staticProps.cropper}
      image={image && image.preview}
      ref={refEditor}
      scale={scale}
    />
    <Slider
      {...staticProps.slider}
      value={scale}
      sliderStyle={styles.slider}
      onChange={onScaleChange}
    />
  </div>
);

ImageCropperRenderer.propTypes = {
  image: T.shape({
    preview: T.string.isRequired,
  }),
  scale: T.number,
  refEditor: T.func.isRequired,
  onScaleChange: T.func.isRequired,
};

ImageCropperRenderer.defaultProps = {
  image: {},
  scale: 1,
};

export const refEditor = ({ setEditor, resetState }) => editor =>
  setEditor(editor && Object.assign(editor, {
    reset: () => {
      editor.state.image = {}; // eslint-disable-line no-param-reassign
      resetState();
    },
    getDataUrl: () => editor.getImageScaledToCanvas().toDataURL(),
  }));

export const onScaleChange = ({ setScale }) => (e, scale) => {
  e.stopPropagation();
  setScale(scale);
};

export {
  ImageCropperRenderer,
};

const handlers = {
  refEditor,
  onScaleChange,
};

export const withImageCropper = compose(
  withStates(['scale']),
  withHandlers(handlers),
);

export default withImageCropper(ImageCropperRenderer);
