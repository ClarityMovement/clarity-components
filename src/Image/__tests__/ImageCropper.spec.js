/* eslint-env jest */
import React from 'react';
import enzymeToJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import ImageCropper, { ImageCropperRenderer } from '../ImageCropper';

const defaultProps = {
  ...ImageCropperRenderer.defaultProps,
  image: { preview: 'http://pic.example.com/200/200' },
};

describe('<ImageCropper {...props} />: elements tree', () => {
  test('with default props', () => {
    const wrapper = shallow(<ImageCropper {...defaultProps} />).until(ImageCropperRenderer);
    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });

  test('with scale controlled', () => {
    const props = {
      ...defaultProps,
      scale: 2,
    };
    const wrapper = shallow(<ImageCropper {...props} />).until(ImageCropperRenderer);
    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });
});
