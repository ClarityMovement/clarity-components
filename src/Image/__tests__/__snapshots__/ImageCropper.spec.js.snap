// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageCropper {...props} />: elements tree with default props 1`] = `
<div>
  <AvatarEditor
    border={34}
    borderRadius={0}
    color={
      Array [
        222,
        222,
        222,
        0.6,
      ]
    }
    crossOrigin="anonymous"
    height={200}
    image="http://pic.example.com/200/200"
    onDropFile={[Function]}
    onImageChange={[Function]}
    onImageReady={[Function]}
    onLoadFailure={[Function]}
    onLoadSuccess={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onPositionChange={[Function]}
    rotate={0}
    scale={1}
    style={Object {}}
    width={200}
  />
  <Slider
    axis="x"
    disableFocusRipple={false}
    disabled={false}
    max={2}
    min={1}
    onChange={[Function]}
    required={true}
    sliderStyle={
      Object {
        "margin": "5px 0",
      }
    }
    step={0.01}
    style={Object {}}
    value={1}
  />
</div>
`;

exports[`<ImageCropper {...props} />: elements tree with scale controlled 1`] = `
<div>
  <AvatarEditor
    border={34}
    borderRadius={0}
    color={
      Array [
        222,
        222,
        222,
        0.6,
      ]
    }
    crossOrigin="anonymous"
    height={200}
    image="http://pic.example.com/200/200"
    onDropFile={[Function]}
    onImageChange={[Function]}
    onImageReady={[Function]}
    onLoadFailure={[Function]}
    onLoadSuccess={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onPositionChange={[Function]}
    rotate={0}
    scale={2}
    style={Object {}}
    width={200}
  />
  <Slider
    axis="x"
    disableFocusRipple={false}
    disabled={false}
    max={2}
    min={1}
    onChange={[Function]}
    required={true}
    sliderStyle={
      Object {
        "margin": "5px 0",
      }
    }
    step={0.01}
    style={Object {}}
    value={2}
  />
</div>
`;
